storage/gcs/src/main/java/io/aiven/kafka/tieredstorage/storage/gcs/GcsStorageConfig.java

```diff
@@ -0,0 +1,179 @@
+                ConfigDef.Importance.MEDIUM,
+                GCS_RESUMABLE_UPLOAD_CHUNK_SIZE_DOC
+            )
+            .define(
+                GCP_CREDENTIALS_JSON_CONFIG,
+                ConfigDef.Type.PASSWORD,
+                null,
+                new NonEmptyPassword(),
+                ConfigDef.Importance.MEDIUM,
+                GCP_CREDENTIALS_JSON_DOC)
+            .define(
+                GCP_CREDENTIALS_PATH_CONFIG,
+                ConfigDef.Type.STRING,
+                null,
+                new ConfigDef.NonEmptyString(),
+                ConfigDef.Importance.MEDIUM,
+                GCP_CREDENTIALS_PATH_DOC)
+            .define(
+                GCP_CREDENTIALS_DEFAULT_CONFIG,
+                ConfigDef.Type.BOOLEAN,
+                true,
+                null,
+                ConfigDef.Importance.MEDIUM,
+                GCP_CREDENTIALS_DEFAULT_DOC);
+    }
+
+    public GcsStorageConfig(final Map<String, ?> props) {
+        super(CONFIG, props);
+        validate();
+    }
+
+    private void validate() {
+        if (getBoolean(GCP_CREDENTIALS_DEFAULT_CONFIG)
+            && getPassword(GCP_CREDENTIALS_JSON_CONFIG) != null) {
+            throw new ConfigException(GCP_CREDENTIALS_JSON_CONFIG
+                + " and "
+                + GCP_CREDENTIALS_DEFAULT_CONFIG
+                + " cannot be set together");
+        }
+
+        if (getBoolean(GCP_CREDENTIALS_DEFAULT_CONFIG)
+            && getString(GCP_CREDENTIALS_PATH_CONFIG) != null) {
+            throw new ConfigException(GCP_CREDENTIALS_PATH_CONFIG
+                + " and "
+                + GCP_CREDENTIALS_DEFAULT_CONFIG
+                + " cannot be set together");
+        }
+
+        if (getPassword(GCP_CREDENTIALS_JSON_CONFIG) != null
+            && getString(GCP_CREDENTIALS_PATH_CONFIG) != null) {
+            throw new ConfigException(GCP_CREDENTIALS_JSON_CONFIG
+                + " and "
+                + GCP_CREDENTIALS_PATH_CONFIG
+                + " cannot be set together");
+        }
+    }
+
+    String bucketName() {
+        return getString(GCS_BUCKET_NAME_CONFIG);
+    }
+
+    String endpointUrl() {
+        return getString(GCS_ENDPOINT_URL_CONFIG);
+    }
+
+    Integer resumableUploadChunkSize() {
+        return getInt(GCS_RESUMABLE_UPLOAD_CHUNK_SIZE_CONFIG);
+    }
+
+    Credentials credentials() {
+        final boolean defaultCredentials = getBoolean(GCP_CREDENTIALS_DEFAULT_CONFIG);
+        final Password credentialsJsonPwd = getPassword(GCP_CREDENTIALS_JSON_CONFIG);
+        final String credentialsJson = credentialsJsonPwd == null ? null : credentialsJsonPwd.value();
+        final String credentialsPath = getString(GCP_CREDENTIALS_PATH_CONFIG);
+        try {
+            return CredentialsBuilder.build(defaultCredentials, credentialsJson, credentialsPath);
+        } catch (final IOException e) {
+            throw new ConfigException("Failed to create GCS credentials: " + e.getMessage());
```

Comment:
any reason not to wrap the exception?
```suggestion
            throw new ConfigException("Failed to create GCS credentials", e);
```

Reply:
That's because a `(String, Throwable)` constructor is not defined for `ConfigException`. There's one in the parent class, but it won't be called.

Reply:
Oh, good to know. Thanks for sharing.